Qcnn example

ihpcdingwj opened the pull request:
Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

scarrazza commented:
What is the difference between this and PR #695?

jf-kong commented:
> What is the difference between this and PR #695?

This PR is for an example (in the folder examples/qcnn_classifier) using the QCNN model based on PR #729. This is based on a previous suggestion by @MatteoRobbiati to have an extended example/tutorial to illustrate the use of the QCNN model. PR #695 is an outdated PR (prior to #729) that has since been closed.

MatteoRobbiati commented on examples/qcnn_classifier/README.md:
```diff
@@ -0,0 +1,109 @@
+# Quantum Convolutional Neural Network Classifier
+
+Code at: [https://github.com/qiboteam/qibo/tree/master/examples/qcnn_classifier]
+(https://github.com/qiboteam/qibo/tree/master/examples/qcnn_classifier).
```
Remove the space between `[...]` and `(...)` in order to make te link work.

MatteoRobbiati commented on examples/qcnn_classifier/README.md:
```diff
@@ -0,0 +1,109 @@
+(https://github.com/qiboteam/qibo/tree/master/examples/qcnn_classifier).
+
+## Problem overview
+This tutorial implements a simple Quantum Convolutional Neural Network (QCNN), which is a translationally invariant algorithm analogous to the classical convolutional neural network. This example demonstrates the use of the QCNN as a quantum classifier, which attempts to classify ground states of a translationally invariant quantum system, the transverse field Ising model, based on whether they a ...
```
```suggestion
This tutorial implements a simple [Quantum Convolutional Neural Network](https://www.nature.com/articles/s41567-019-0648-8) (QCNN), which is a translationally invariant algorithm analogous to the classical [convolutional neural network](https://proceedings.neurips.cc/paper_files/paper/2012/file/c399862d3b9d6b76c8436e924a68c45b-Paper.pdf). This example demonstrates the use of the QCNN as a quantum classifier, which attempts to classify ground states of a translationally invariant quantum system, the transverse field Ising model, based on whether they are in the ordered or disordered phase. The (randomized) statevector data provided are those of a 4-qubit system. Accompanying each state is a label: +1 (ordered phase) or -1 (ordered phase).

```

MatteoRobbiati replied:
I think one of the two (ordered phase) should be (unordered phase).

MatteoRobbiati commented on examples/qcnn_classifier/README.md:
```diff
@@ -0,0 +1,109 @@
+![RT](images/RT.PNG)
+
+## How to use the QCNN class
+For more details on the QuantumCNN class, please refer to the documentation. Here we recall some of the necessary arguments when instantiating a QuantumCNN object:
```
Please add link to the documentation when it is available (where you mention the docs).

MatteoRobbiati reviewed (changes requested):
Thank you very much for this, it is very useful for a better understanding of the model you implemented. 

Let me add some suggestions:

1. this is not the first example in `qibo` based on an IPython implementation. The [qPDF](https://github.com/qiboteam/qibo/tree/qcnn-example/examples/qPDF)  is another one. Personally, I think the notebook approach is very useful in terms of visualization of the problem (e.g. in the documentation), but I think [this](https://github.com/qiboteam/qibo/tree/master/examples) folder starts to be a bit messy. 
According to this, I purpose two solutions: the first one is to remove the implementation part from `README.md` (I mean from `## How to use the QCNN class`), which is not needed since already shown into the notebook. On this way this example will follow the line of the qPDF one. The second is to change this example into a Python script as is done for all the examples in `qibo` and to move this notebook to the [tutorials](https://github.com/qiboteam/tutorials) repository, in order to be used into the documentation. 
I prefer this second option, since notebooks are useful tools but very difficult to handle with several hands.
2. please select only one of the two notebooks and fix it: there are some unused packages and modules there (e.g. `gates, numpy` etc.) and the `qcnn` package must be loaded properly (`from qibo.models.qcnn import QuantumCNN`); 
3. add your example [here](https://github.com/qiboteam/qibo/blob/master/examples/README.md).

Some small suggestions follow.

MatteoRobbiati commented:
What is the status here @ihpcdingwj?

Let me summarize some remaining tasks:

- [ ] remove one of the two notebooks. You have two versions of the same example;
- [ ] fix the imports in the notebook: some packages are unused;
- [ ] fix the `qcnn` import by setting `from qibo.models.qcnn import QuantumCNN`;
- [ ] please follow the [general convention](https://peps.python.org/pep-0008/#imports): first Standard library imports, then related third party imports, then
local application/library specific imports.
- [ ] add your example [here](https://github.com/qiboteam/qibo/blob/master/examples/README.md);
- [ ] since we decided to keep the jupyter notebook form, please remove the part of `README.md` which is at this point reduntant: the `##How to use the QCNN class` is not needed if already described in the notebook.

MatteoRobbiati reviewed (approved):
I did some small fix to the notebook (like passing `circuit.draw()` to the `print` function, etc),
fixed the link2path in the `README.md` and changed the name of the notebook into `qcnn_demo.ipynb`.

The last suggestion is to explicitly write in the `README.md` that `scikit-learn` is needed to run your example. It should be better to link the [official installation page](https://scikit-learn.org/stable/install.html).

I also suggest to rebase this branch on the master.

ihpcdingwj commented:
> What is the status here @ihpcdingwj?
> 
> Let me summarize some remaining tasks:
> 
> * [x]  remove one of the two notebooks. You have two versions of the same example;
> * [x]  fix the imports in the notebook: some packages are unused;
> * [x]  fix the `qcnn` import by setting `from qibo.models.qcnn import QuantumCNN`;
> * [x]  please follow the [general convention](https://peps.python.org/pep-0008/#imports): first Standard library imports, then related third party imports, then
>   local application/library specific imports.
> * [x]  add your example [here](https://github.com/qiboteam/qibo/blob/master/examples/README.md);
> * [x]  since we decided to keep the jupyter notebook form, please remove the part of `README.md` which is at this point reduntant: the `##How to use the QCNN class` is not needed if already described in the notebook.

@MatteoRobbiati We have made all the suggested edits. Please let us know if anything else to address. Thanks.

ihpcdingwj commented:
> I did some small fix to the notebook (like passing `circuit.draw()` to the `print` function, etc), fixed the link2path in the `README.md` and changed the name of the notebook into `qcnn_demo.ipynb`.
> 
> The last suggestion is to explicitly write in the `README.md` that `scikit-learn` is needed to run your example. It should be better to link the [official installation page](https://scikit-learn.org/stable/install.html).
> 
> I also suggest to rebase this branch on the master.

Thanks @MatteoRobbiati for the editing. We have made the changes suggested. Could you also clarify how do we proceed with "rebase"? 

> I did some small fix to the notebook (like passing `circuit.draw()` to the `print` function, etc), fixed the link2path in the `README.md` and changed the name of the notebook into `qcnn_demo.ipynb`.
> 
> The last suggestion is to explicitly write in the `README.md` that `scikit-learn` is needed to run your example. It should be better to link the [official installation page](https://scikit-learn.org/stable/install.html).
> 
> I also suggest to rebase this branch on the master.

Thanks @MatteoRobbiati for the editing. We have made the changes suggested. Could you also clarify how do we proceed with "rebase"?

MatteoRobbiati commented:
> Thanks @MatteoRobbiati for the editing. We have made the changes suggested. Could you also clarify how do we proceed with "rebase"?

Thank you for the modification.
With rebase I mean [this git operation](https://git-scm.com/docs/git-rebase), but I think in this case we don't have conflicts.

jf-kong commented:
> > Thanks @MatteoRobbiati for the editing. We have made the changes suggested. Could you also clarify how do we proceed with "rebase"?
> 
> Thank you for the modification. With rebase I mean [this git operation](https://git-scm.com/docs/git-rebase), but I think in this case we don't have conflicts.

Thanks @MatteoRobbiati. Have rewritten history :)
